### ISSUE TITLE
Horovod using tensor fusion results in unstable gradients

### ISSUE THREAD

**Issue by Richie-yan** [horovod/horovod#1548]
**Environment:**
1. Framework: PyTorch 
2. Framework version: 1.3.0 
3. Horovod version:0.18.2
4. MPI version: ompi-4.0.0
5. CUDA version:9.0
6. NCCL version:2.3.7
7. Python version:3.6.8
8. OS and version:c7 
9. GCC version:4.9

When I use horovod to run the mnist example of pytorch, the gradient of the network parameters is saved after optimizer.step (), and it is found that when the tensor fusion function of horovod is turned on, the gradient values at the same step of each experiment was not Stable, with inconsistent values. I can guarantee that there is no randomness in each experiment, because compared to the training of multiple GPUs using pytorch's DistributedDataParallel, the gradient results are stable, and the saved gradient results are consistent, but horovod's gradient of each experiment Inconsistent, why is the gradient unstable?

**Comment by huoliquankai7** [comment]
Here are my test cases：convenient to reproduce this problem
Running the following test multiple times, you will find that the md5 of the saved gradient is inconsistent
horovodrun -np 3 -H localhost:3 python py_mn.py

#py_mn.py
from __future__ import print_function
import argparse
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
import torch.utils.data.distributed
import horovod.torch as hvd
import numpy as np

# Training settings
parser = argparse.ArgumentParser(description='PyTorch MNIST Example')
parser.add_argument('--batch-size', type=int, default=64, metavar='N',
                    help='input batch size for training (default: 64)')
parser.add_argument('--test-batch-size', type=int, default=1000, metavar='N',
                    help='input batch size for testing (default: 1000)')
parser.add_argument('--epochs', type=int, default=10, metavar='N',
                    help='number of epochs to train (default: 10)')
parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                    help='learning rate (default: 0.01)')
parser.add_argument('--momentum', type=float, default=0.5, metavar='M',
                    help='SGD momentum (default: 0.5)')
parser.add_argument('--no-cuda', action='store_true', default=False,
                    help='disables CUDA training')
parser.add_argument('--seed', type=int, default=42, metavar='S',
                    help='random seed (default: 42)')
parser.add_argument('--log-interval', type=int, default=10, metavar='N',
                    help='how many batches to wait before logging training status')
parser.add_argument('--fp16-allreduce', action='store_true', default=False,
                    help='use fp16 compression during allreduce')
args = parser.parse_args()

# Horovod: initialize library.
hvd.init()
torch.manual_seed(611+hvd.local_rank())

torch.cuda.set_device(hvd.local_rank())
torch.cuda.manual_seed(611+hvd.local_rank())

# Horovod: limit # of CPU threads to be used per worker.
torch.set_num_threads(1)

kwargs = {'num_workers': 1, 'pin_memory': True}
train_dataset = \
    datasets.MNIST('data-%d' % hvd.rank(), train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ]))
# Horovod: use DistributedSampler to partition the training data.
train_sampler = torch.utils.data.distributed.DistributedSampler(
    train_dataset, num_replicas=hvd.size(), rank=hvd.rank())
train_loader = torch.utils.data.DataLoader(
    train_dataset, batch_size=args.batch_size, sampler=train_sampler, **kwargs)

test_dataset = \
    datasets.MNIST('data-%d' % hvd.rank(), train=False, transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
    ]))
# Horovod: use DistributedSampler to partition the test data.
test_sampler = torch.utils.data.distributed.DistributedSampler(
    test_dataset, num_replicas=hvd.size(), rank=hvd.rank())
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=args.test_batch_size,
                                          sampler=test_sampler, **kwargs)

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x)

model = Net()
model.cuda()

# Horovod: scale learning rate by the number of GPUs.
#optimizer = optim.SGD(model.parameters(), lr=args.lr * hvd.size(),
#                      momentum=args.momentum)
optimizer = optim.SGD(model.parameters(), lr=0.03,
                      momentum=0.5)

# Horovod: broadcast parameters & optimizer state.
hvd.broadcast_parameters(model.state_dict(), root_rank=0)
hvd.broadcast_optimizer_state(optimizer, root_rank=0)

# Horovod: (optional) compression algorithm.
compression = hvd.Compression.fp16 if args.fp16_allreduce else hvd.Compression.none

# Horovod: wrap optimizer with DistributedOptimizer.
optimizer = hvd.DistributedOptimizer(optimizer,
                                     named_parameters=model.named_parameters(),
                                     compression=compression)
def train(epoch):
    model.train()
    train_sampler.set_epoch(epoch)
    for i in range(20):
        data = torch.randn(64, 1, 28, 28)
        data = data.cuda()
        target = torch.from_numpy(np.load("../label.npy")).cuda()#note:Load a fixed label here, the 
                                                              #label can be directly generated by hortord's pytorch_mnist.py
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if i==15:
            for index,param in enumerate(model.parameters()):
                if hasattr(param, "grad"):
                    np.save("%s_%s.npy"%(hvd.local_rank(),index), param.grad.cpu().numpy())
                else:
                    print("this object have no grad")
            break    
train(1)

**Comment by romerojosh** [comment]
One possible reason for this is that the tensor fusion in Horovod is dynamic and the sets of gradients tensors that are fused from run to run (or even training step to training step) are not guaranteed to be the same. As a result, the NCCL allreduce operation may perform reductions with different summation ordering, resulting in small numerical differences in the gradients.

**Comment by Richie-yan** [comment]
@romerojosh 
After testing, tensor fusion in Horovod is dynamic.
What I am more confused is why the different sets of gradients tensors that are fused will cause gradient inconsistencies? Is there a problem with the nccl segmentation? Why does nccl's different summation ordering cause small numerical differences in the gradients?

**Comment by romerojosh** [comment]
The buffer submitted to NCCL will be partitioned into chunks, with different chunks having their summation start on a different GPU. The dynamic tensor fusion generates different configurations of the fused buffer, causing NCCL to vary which GPU is the first to sum a given entry. Summation ordering causing small numerical differences isn't a NCCL specific issue, but is a general phenomenon that occurs with floating-point summation. 
With that being said, this type of non-determinism in the gradients shouldn't have a big impact on your training convergence.

**Comment by Richie-yan** [comment]
@romerojosh Thank you for your reply.
 I probably understand what you mean. When using nccl for allreduce, for the same part of the data, dynamic fusion will eventually cause the order of summation to be different. For floating point numbers, different summation order will result in different results.
